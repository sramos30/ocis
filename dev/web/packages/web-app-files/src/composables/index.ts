export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './sort'

declare module 'vue/types/vue' {
  interface Vue {
    [key: string]: any
  }
}
