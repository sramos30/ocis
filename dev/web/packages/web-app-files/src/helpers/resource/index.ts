export * from './asset'
export * from './filter'
export * from './privatePreviewBlob'
export * from './publicPreviewUrl'
export * from './resource'
export * from './sameResource'
