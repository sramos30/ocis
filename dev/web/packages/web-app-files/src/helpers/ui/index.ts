export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
