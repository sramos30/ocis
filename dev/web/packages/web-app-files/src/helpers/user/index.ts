export * from './avatarUrl'
export * from './types'
