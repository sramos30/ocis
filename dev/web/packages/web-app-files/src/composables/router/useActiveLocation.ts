import { ref, Ref, watch } from '@vue/composition-api'
import { useRoute, useRouter } from 'web-pkg/src/composables'
import { ActiveRouteDirectorFunc } from '../../router'

/**
 * watches the current route and re-evaluates the provided active location director
 * on each route name update.
 *
 * @param director
 * @param comparatives
 */
export const useActiveLocation = <T extends string>(
  director: ActiveRouteDirectorFunc<T>,
  ...comparatives: T[]
): Ref<boolean> => {
  const value = ref(false)
  const router = useRouter()
  const currentRoute = useRoute()
  watch(
    currentRoute,
    () => {
      value.value = director(router, ...comparatives)
    },
    { immediate: true }
  )
  return value
}
