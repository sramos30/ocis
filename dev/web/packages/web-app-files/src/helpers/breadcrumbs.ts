import { bus } from 'web-pkg/src/instance'

export interface BreadcrumbItem {
  text: string
  to?: string
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentPath: string,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentPath)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/')
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => bus.publish('app.files.list.load')
    }
  ]
}
