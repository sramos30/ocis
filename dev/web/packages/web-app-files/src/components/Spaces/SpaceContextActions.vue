<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :items="items" />
    <quota-modal
      v-if="quotaModalIsOpen"
      :cancel="closeQuotaModal"
      :space="quotaModalSelectedSpace"
    />
    <readme-content-modal
      v-if="readmeContentModalIsOpen"
      :cancel="closeReadmeContentModal"
      :space="items[0]"
    />
    <input
      id="space-image-upload-input"
      ref="spaceImageInput"
      type="file"
      name="file"
      multiple
      tabindex="-1"
      accept="image/*"
      @change="$_uploadImage_uploadImageSpace"
    />
  </div>
</template>

<script>
import ContextActionMenu from '../ContextActionMenu.vue'
import QuotaModal from './QuotaModal.vue'
import ReadmeContentModal from './ReadmeContentModal.vue'

import Delete from '../../mixins/spaces/actions/delete'
import Rename from '../../mixins/spaces/actions/rename'
import Restore from '../../mixins/spaces/actions/restore'
import ShowDetails from '../../mixins/spaces/actions/showDetails'
import EditDescription from '../../mixins/spaces/actions/editDescription'
import EditQuota from '../../mixins/spaces/actions/editQuota'
import DeletedFiles from '../../mixins/spaces/actions/deletedFiles'
import Disable from '../../mixins/spaces/actions/disable'
import ShowMembers from '../../mixins/spaces/actions/showMembers'
import UploadImage from '../../mixins/spaces/actions/uploadImage'
import EditReadmeContent from '../../mixins/spaces/actions/editReadmeContent'
import { isLocationSpacesActive } from '../../router'

export default {
  name: 'SpaceContextActions',
  components: { ContextActionMenu, QuotaModal, ReadmeContentModal },
  mixins: [
    Rename,
    Delete,
    EditDescription,
    EditQuota,
    DeletedFiles,
    Disable,
    ShowDetails,
    ShowMembers,
    Restore,
    UploadImage,
    EditReadmeContent
  ],

  props: {
    items: {
      type: Array,
      required: true
    }
  },

  computed: {
    quotaModalSelectedSpace() {
      return this.$data.$_editQuota_selectedSpace
    },
    quotaModalIsOpen() {
      return this.$data.$_editQuota_modalOpen
    },
    readmeContentModalIsOpen() {
      return this.$data.$_editReadmeContent_modalOpen
    },
    menuSections() {
      const sections = []

      if (this.menuItemsMembers.length) {
        sections.push({
          name: 'members',
          items: this.menuItemsMembers
        })
      }
      if (this.menuItemsPrimaryActions.length) {
        sections.push({
          name: 'primaryActions',
          items: this.menuItemsPrimaryActions
        })
      }
      if (this.menuItemsSecondaryActions.length) {
        sections.push({
          name: 'secondaryActions',
          items: this.menuItemsSecondaryActions
        })
      }
      if (this.menuItemsTrashBin.length) {
        sections.push({
          name: 'trashBin',
          items: this.menuItemsTrashBin
        })
      }
      if (this.menuItemsSidebar.length) {
        sections.push({
          name: 'sidebar',
          items: this.menuItemsSidebar
        })
      }

      return sections
    },

    filterParams() {
      return {
        resources: this.items
      }
    },

    menuItemsMembers() {
      const fileHandlers = [...this.$_showMembers_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsPrimaryActions() {
      const fileHandlers = [
        ...this.$_rename_items,
        ...this.$_editDescription_items,
        ...this.$_uploadImage_items
      ]

      if (
        isLocationSpacesActive(this.$router, 'files-spaces-project') &&
        this.$route.params.storageId
      ) {
        fileHandlers.splice(2, 0, ...this.$_editReadmeContent_items)
      }
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsSecondaryActions() {
      const fileHandlers = [
        ...this.$_editQuota_items,
        ...this.$_disable_items,
        ...this.$_restore_items,
        ...this.$_delete_items
      ]

      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsTrashBin() {
      const fileHandlers = [...this.$_deletedFiles_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsSidebar() {
      const fileHandlers = [...this.$_showDetails_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    }
  },
  methods: {
    closeQuotaModal() {
      this.$_editQuota_closeModal()
    },
    closeReadmeContentModal() {
      this.$_editReadmeContent_closeModal()
    }
  }
}
</script>

<style lang="scss">
#space-image-upload-input {
  position: absolute;
  left: -99999px;
}
</style>
