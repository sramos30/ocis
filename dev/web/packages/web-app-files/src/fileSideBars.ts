import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import FileShares from './components/SideBar/Shares/FileShares.vue'
import FileLinks from './components/SideBar/Links/FileLinks.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from './components/SideBar/Details/SpaceDetails.vue'
import SpaceShares from './components/SideBar/Shares/SpaceShares.vue'
import { isLocationSpacesActive, isLocationTrashActive, isLocationPublicActive } from './router'
import { spaceRoleEditor, spaceRoleManager } from './helpers/share'
import { Panel } from '../../web-pkg/src/components/sidebar'

import { Resource } from './helpers/resource'
import { User } from './helpers/user'
import Router from 'vue-router'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  highlightedFile,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  highlightedFile: Resource
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ rootFolder, highlightedFile }): Panel => ({
    app: 'no-selection-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return rootFolder && highlightedFile?.type !== 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'details-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default:
      !isLocationTrashActive(router, 'files-trash-personal') &&
      !isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return (
        !isLocationTrashActive(router, 'files-trash-personal') &&
        !isLocationTrashActive(router, 'files-trash-spaces-project') &&
        !multipleSelection &&
        !rootFolder
      )
    }
  }),
  ({ multipleSelection, rootFolder }) => ({
    app: 'details-multiple-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    default: () => true,
    get enabled() {
      return multipleSelection && !rootFolder
    }
  }),
  ({ router, highlightedFile }) => ({
    app: 'details-space-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: isLocationSpacesActive(router, 'files-spaces-projects'),
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default:
      isLocationTrashActive(router, 'files-trash-personal') ||
      isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return !multipleSelection && !rootFolder
    }
  }),
  ({ highlightedFile, user }) => ({
    app: 'space-actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (highlightedFile?.type !== 'space') {
        return false
      }
      return [
        ...highlightedFile.spaceRoles[spaceRoleManager.name],
        ...highlightedFile.spaceRoles[spaceRoleEditor.name]
      ].includes(user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder }) => ({
    app: 'sharing-item',
    icon: 'group',
    title: $gettext('People'),
    component: FileShares,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ highlightedFile }) => ({
    app: 'space-share-item',
    icon: 'group',
    title: $gettext('Members'),
    component: SpaceShares,
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'links-item',
    icon: 'link',
    title: $gettext('Links'),
    component: FileLinks,
    get enabled() {
      if (multipleSelection || (rootFolder && highlightedFile?.type !== 'space')) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.public.enabled
      }
      return false
    }
  }),
  ({ capabilities, highlightedFile, router, multipleSelection, rootFolder }) => ({
    app: 'versions-item',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }
      return !!capabilities.core && highlightedFile && highlightedFile.type !== 'folder'
    }
  })
]

export default panelGenerators
