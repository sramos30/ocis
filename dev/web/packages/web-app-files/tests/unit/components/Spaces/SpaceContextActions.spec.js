import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import { createLocalVue, mount } from '@vue/test-utils'
import SpaceContextActions from '../../../../src/components/Spaces/SpaceContextActions.vue'
import GetTextPlugin from 'vue-gettext'
import { createLocationSpaces } from '../../../../src/router'
import { buildSpace } from '../../../../src/helpers/resources'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const spaceMock = {
  id: '1',
  root: { permissions: [{ roles: ['manager'], grantedTo: [{ user: { id: 1 } }] }] }
}

describe('SpaceContextActions', () => {
  describe('action handlers', () => {
    it('renders actions that are always available: "Members", "Deleted files", "Details"', () => {
      const wrapper = getWrapper(buildSpace(spaceMock))
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function getWrapper(space) {
  return mount(SpaceContextActions, {
    localVue,
    store: createStore(),
    mocks: {
      $router: {
        currentRoute: createLocationSpaces('some-route'),
        resolve: (r) => {
          return { href: r.name }
        }
      }
    },
    propsData: {
      items: [space]
    }
  })
}

function createStore() {
  return new Vuex.Store({})
}
