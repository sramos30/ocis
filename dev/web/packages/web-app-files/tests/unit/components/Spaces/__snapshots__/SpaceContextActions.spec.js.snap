// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceContextActions action handlers renders actions that are always available: "Members", "Deleted files", "Details" 1`] = `
<div>
  <div id="oc-files-context-menu">
    <ul id="oc-files-context-actions-members" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-files-context-actions-border">
      <li class="oc-files-context-action oc-px-s oc-rounded"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Members</span>
          <!---->
        </button></li>
    </ul>
    <ul id="oc-files-context-actions-trashBin" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-pt-s oc-files-context-actions-border">
      <li class="oc-files-context-action oc-px-s oc-rounded"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-delete-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Deleted files</span>
          <!---->
        </button></li>
    </ul>
    <ul id="oc-files-context-actions-sidebar" class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pt-s">
      <li class="oc-files-context-action oc-px-s oc-rounded"><button type="button" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" data-testid="action-handler"><span class="oc-icon oc-icon-m oc-icon-passive" data-testid="action-icon"><!----></span> <span data-testid="action-label" class="oc-files-context-action-label">Details</span>
          <!---->
        </button></li>
    </ul>
  </div>
  <!---->
  <!----> <input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
</div>
`;
